File: interviews/WorldQuant/p4.cpp
#include <vector>
using namespace std;


int BIT[100005], n;
void update(int x, int delta)
{     
      for(; x <= n; x += x&-x){
        BIT[x] += delta;
      }
}
int query(int x)
{
     int sum = 0;
     for(; x > 0; x -= x&-x)
        sum += BIT[x];
     return sum;
} 
 
long getCount( long number, int i){
    int left = query(number-1);
    int right = i - query(number);
    return left < right ? 2*left+1 : 2*right+1; 
}

long minimumOperations(vector<int> numbers) {
    long ans=0;
    n = 0;
    for(int i=0;i<numbers.size();i++){
         n = max(n, numbers[i]);
    }
    for(int i=0; i<numbers.size(); i++){
        ans += getCount(numbers[i], i);
        update(numbers[i],1);
    }
    return ans;
}

File: interviews/WorldQuant/p4.py
def minimumOperations(numbers):
    # Write your code here
    # n = len(numbers)
    # total_moves = 0
    # for i in range(len(numbers)):
    #     cnt_left = 0
    #     cnt_right = 0
    #     for j in range(i):
    #         if numbers[i] > numbers[j]:
    #             cnt_left += 1
    #         elif numbers[i] < numbers[j]:
    #             cnt_right += 1
    #     total_moves += min(cnt_left, cnt_right)
    # return n + 2 * total_moves
    def getCount(arr, num):
        if len(arr) == 0 or len(arr) == 1 or max(arr) == num:
            return 1
        else:
            left = 0
            right = 0
            for i in range(len(arr)):
                if num < arr[i]:
                    left = left + 1
                elif num > arr[i]:
                    right = right + 1
            return right * 2 + 1 if right <= left else left * 2 + 1

    arr = []
    tot = 0
    for i in range(len(numbers)):
        tot = tot + getCount(arr, numbers[i])
        arr.append(numbers[i])
    return tot

if __name__ == '__main__':
    print(minimumOperations([10, 6, 2, 3, 7, 1, 2]))
